### PR TITLE
support sidecar update

### PR DESCRIPTION
**Pull request by coldTea214**
### Ⅰ. Describe what this PR does

support sidecar update. feature list:

* support maxUnavailable=1 update strategy
* only support image update currently
* support pause, unpause

### Ⅲ. List the added test cases (unit test/integration test) if any, please explain if no tests are needed.

pkg/webhook/default_server/pod/mutating/pod_mutating_test.go
pkg/webhook/default_server/sidecarset/mutating/sidecarset_mutating_test.go
pkg/controller/sidecarset/sidecarset_controller_test.go

### Ⅳ. Describe how to verify it

create a sidecarset and some matched pods or high level workloads, then update image in sidecarset. 

* check whether sidecars in pods are updated
* during the update process, check whether sidecars are updated one by one(maxUnavailable=1)
* during the update process, use kubectl to update pause field, check whether update process is paused/unpaused 

### Ⅴ. Special notes for reviews

add paused field, which is similar to paused in deployment semantically

### CODE REVIEW

File: docs/concepts/sidecarSet/README.md [hunk 1]
```diff
@@ -98,6 +98,17 @@ status:
   readyPods: 1
   updatedPods: 1
 ```
+
+### Update a SidecarSet
+
+Use ```kubectl edit``` to modify sidecarSet image from "centos:6.7" to "centos:6.8", then you will see matched pods will be updated in place one by one(we only support maxUnavailable=1 for now).
```
Comment (Fei-Guo) [inline docs/concepts/sidecarSet/README.md]: Can you provide a full command line here?

File: docs/concepts/sidecarSet/README.md [hunk 2]
```diff
@@ -98,6 +98,17 @@ status:
...
+
+You could use ```kubectl patch sidecarset test-sidecarset --type merge -p '{"spec":{"paused":true}}'``` to pause the update procedure.
+
+If user modify fields other than image in sidecarSet, pod won't get updated until pod is recreated by workload(eg. deploy), which we called "lazy update".
```
Comment (Fei-Guo) [inline docs/concepts/sidecarSet/README.md]: modify->modifies 
eg. -> e.g., 
deploy->deployment (or StatefulSet)

File: pkg/webhook/default_server/sidecarset/mutating/hash.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
+package mutating
```
Comment (Fei-Guo) [inline pkg/webhook/default_server/sidecarset/mutating/hash.go]: Can you add copyright header here?

File: pkg/controller/sidecarset/sidecarset_utils.go [hunk 4]
```diff
@@ -73,20 +83,21 @@ func (r *ReconcileSidecarSet) updateSidecarSetStatus(sidecarSet *appsv1alpha1.Si
...
 
 		key := types.NamespacedName{
-			Name: sidecarSet.Name,
+			Name: sidecarSetClone.Name,
 		}
 
-		if err := r.Get(context.TODO(), key, sidecarSet); err != nil {
-			klog.Errorf("error getting updated service %s from client", sidecarSet.Name)
+		if err := r.Get(context.TODO(), key, sidecarSetClone); err != nil {
+			klog.Errorf("error getting updated sidecarset %s from client", sidecarSetClone.Name)
```
Comment (Fei-Guo) [inline pkg/controller/sidecarset/sidecarset_utils.go]: Since input sidecarSet will be used in the later, should we update sidecarSet with the latest version  here?
} else {
    sidecarSet = sidecarSetClone.DeepCopy() 
}

File: pkg/controller/sidecarset/sidecarset_controller.go [hunk 5]
```diff
@@ -131,7 +130,29 @@ func (r *ReconcileSidecarSet) Reconcile(request reconcile.Request) (reconcile.Re
 		return reconcile.Result{}, err
 	}
 
-	// update sidecarset status
 	err = r.updateSidecarSetStatus(sidecarSet, status)
```
Comment (Fei-Guo) [inline pkg/controller/sidecarset/sidecarset_controller.go]: Should we move the status update in the end of the method and add stats about how many pods have been upgraded in the new version?

File: pkg/controller/sidecarset/sidecarset_utils.go [hunk 6]
```diff
@@ -101,3 +112,182 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
...
+	}
+	return filteredPods, nil
+}
+
+func hashChanged(podHashs, targetName, targetHash string) (bool, error) {
+	sidecarSetHash := make(map[string]string)
+	if err := json.Unmarshal([]byte(podHashs), &sidecarSetHash); err != nil {
+		return false, err
+	}
+	return sidecarSetHash[targetName] != targetHash, nil
```
Comment (Fei-Guo) [inline pkg/controller/sidecarset/sidecarset_utils.go]: It is possible that the Pod does not have any annotation. (e.g., A matching Pod created before SetcarSet is created) . So sidecarSetHash can be an empty map and the hashChanged method can return true. We should exclude this kind of Pods from filtered Pods.

File: docs/concepts/sidecarSet/README.md [hunk 7]
```diff
@@ -53,7 +53,7 @@ spec:
       app: nginx
   containers:
   - name: sidecar1
-    image: centos:7
+    image: centos:6.7
```
Comment (xiang90) [inline docs/concepts/sidecarSet/README.md]: why is this changed?

File: pkg/controller/sidecarset/sidecarset_utils.go [hunk 8]
```diff
@@ -101,3 +112,182 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
...
+}
+
+// available definition:
+// 1. image in pod.spec and pod.status is exactly the same
+// 2. pod is ready
+func getUnavailableNumber(sidecarSet *appsv1alpha1.SidecarSet, pods []*corev1.Pod) (int, error) {
+	var unavailableNum int
+	for _, pod := range pods {
+		// in case of informer cache latency
+		podInCacheUpdated := updateCache.isUpdated(
```
Comment (Fei-Guo) [inline pkg/controller/sidecarset/sidecarset_utils.go]: Can you explain more why we need this updateCache?

File: pkg/controller/sidecarset/sidecarset_controller.go [hunk 9]
```diff
@@ -131,7 +130,29 @@ func (r *ReconcileSidecarSet) Reconcile(request reconcile.Request) (reconcile.Re
...
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	// we only support sequential update currently, equals to maxUnavailable = 1
+	if unavailableNum != 0 {
+		klog.V(3).Infof("current unavailable pod number: %v, skip update", unavailableNum)
+		return reconcile.Result{}, nil
+	}
+
+	if sidecarSet.Spec.Paused {
```
Comment (furykerry) [inline pkg/controller/sidecarset/sidecarset_controller.go]: Should we move Pause checking before unavailable pods checking ? after all, if sidecarset is paused, unavailable number is no longer used.

File: docs/concepts/sidecarSet/README.md (hunk 7 above)
Comment (coldTea214) [reply docs/concepts/sidecarSet/README.md]: OCD. :)
in my README.md, it makes me feel better to change from centos:6.7 to centos:6.8 instead of changing from centos:7 to centos:6.8

File: docs/concepts/sidecarSet/README.md (hunk 2 above)
Comment (coldTea214) [reply docs/concepts/sidecarSet/README.md]: I will fix it

File: pkg/controller/sidecarset/sidecarset_utils.go (hunk 8 above)
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_utils.go]: sure. even if controller updated pod in the last syncLoop, the pod might still be considered as not updated (happens when informer cache doen't update in time) next time, which will cause maxUnavailable=0

**Comment by allencloud** [comment]
Please strictly follow the PR description template. Thanks a lot. @coldTea214

File: pkg/controller/sidecarset/sidecarset_utils.go [hunk 10]
```diff
@@ -101,3 +112,182 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
...
+
+		if !isRunningAndReady(pod) {
+			unavailableNum++
+		}
+	}
+	return unavailableNum, nil
+}
+
+func isPodImageConsistent(pod *corev1.Pod) bool {
+	containerSpecImage := make(map[string]string)
```
Comment (furykerry) [inline pkg/controller/sidecarset/sidecarset_utils.go]: make(map[string]string, len(pod.spec.Containers)?

File: pkg/controller/sidecarset/sidecarset_utils.go [hunk 11]
```diff
@@ -101,3 +112,182 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
...
+		}
+
+		return updateErr
+	})
+
+	return err
+}
+
+func updatePodSidecar(sidecarSet *appsv1alpha1.SidecarSet, pod *corev1.Pod) {
+	sidecarImage := make(map[string]string)
```
Comment (furykerry) [inline pkg/controller/sidecarset/sidecarset_utils.go]: make(map[string]string, len(sidecarSet.spec.Containers)?

File: pkg/controller/sidecarset/sidecarset_utils.go [hunk 12]
```diff
@@ -101,3 +112,182 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
 		status.UpdatedPods != sidecarSet.Status.UpdatedPods ||
 		status.ReadyPods != sidecarSet.Status.ReadyPods
 }
+
+// add this cache to avoid be influenced by informer cache latency when controller try to count maxUnavailable
+type updatedPodCache struct {
```
Comment (furykerry) [inline pkg/controller/sidecarset/sidecarset_utils.go]: please consider make updatePodCache an interface and provide a default implementation

File: pkg/controller/sidecarset/sidecarset_utils.go [hunk 13]
```diff
@@ -101,3 +112,182 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
...
+	podSidecarUpdated map[string]string
+}
+
+func (u *updatedPodCache) set(key, hash string) {
+	u.lock.Lock()
+	u.podSidecarUpdated[key] = hash
+	u.lock.Unlock()
+}
+
+func (u *updatedPodCache) reset() {
```
Comment (furykerry) [inline pkg/controller/sidecarset/sidecarset_utils.go]: do we need a delete method if a pod is deleted, otherwise the cache will grow indefinitely

**Comment by coldTea214** [comment]
> * Is it possible to use ControllerRevision to save the sidecar upgrade history and use it to check if it is an in-place upgrade or not instead of using hash? (similar to AdvancedStatefulSet). This can also help user to rollback to previous version.

I don't think so. It's not gonna work to bind a non-namespaced workload to a namespaced controller revision. So is it necessary to define another crd which is similar to controller revision? Maybe we could work on that later.

File: pkg/controller/sidecarset/sidecarset_controller.go (hunk 5 above)
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_controller.go]: in fact, it makes no difference to move it to the end of this function. because we  calculate status from informer cache, which won't get updated until next sync loop.

File: pkg/controller/sidecarset/sidecarset_utils.go (hunk 12 above)
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_utils.go]: it's not a good idea to make updatePodCache into a interface at the first place. because nobody knows what will be added into the implemetation.
it will not be refactored into a interface until:
* methods is stable
* other packages need this cache
otherwise I don't think it's necessary to do this change.

**Comment by coldTea214** [comment]
> please improve update cache interface and separate cache interface and design

no, I don't think it's a good idea to do that

File: pkg/controller/sidecarset/sidecarset_utils.go (hunk 13 above)
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_utils.go]: done, add a removeStaleCache function

File: pkg/controller/sidecarset/sidecarset_utils.go (hunk 6 above)
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_utils.go]: done, add isPodCreatedBeforeSidecarSet function

File: docs/concepts/sidecarSet/README.md (hunk 2 above)
Comment (coldTea214) [reply docs/concepts/sidecarSet/README.md]: done

File: pkg/webhook/default_server/sidecarset/mutating/hash.go (hunk 3 above)
Comment (coldTea214) [reply pkg/webhook/default_server/sidecarset/mutating/hash.go]: done

File: pkg/controller/sidecarset/sidecarset_controller.go (hunk 9 above)
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_controller.go]: done

File: pkg/controller/sidecarset/sidecarset_utils.go (hunk 10 above)
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_utils.go]: done

File: pkg/controller/sidecarset/sidecarset_utils.go (hunk 11 above)
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_utils.go]: done

File: pkg/webhook/default_server/pod/mutating/pod_create_handler.go [hunk 14]
```diff
@@ -92,9 +95,19 @@ func (h *PodCreateHandler) sidecarsetMutatingPod(ctx context.Context, pod *corev
 		}
 		matchNothing = false
 
-		sidecarSetGeneration[sidecarSet.Name] = sidecarSet.Generation
+		hash, err := mutating.SidecarSetHash(&sidecarSet)
```
Comment (Fei-Guo) [inline pkg/webhook/default_server/pod/mutating/pod_create_handler.go]: Can you directly read the hash from sidecarset annotation instead of recomputing it from spec? 
The same for the SidecarsetHashWithoutImage.

File: pkg/controller/sidecarset/sidecarset_utils.go [hunk 15]
```diff
@@ -101,3 +112,218 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
...
+		}
+
+		if !isRunningAndReady(pod) {
+			unavailableNum++
+		}
+	}
+	return unavailableNum, nil
+}
+
+func isPodCreatedBeforeSidecarSet(pod *corev1.Pod) bool {
```
[the remark this answers is not in the document]
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_utils.go]: brilliant! my mistake

File: pkg/webhook/default_server/sidecarset/mutating/hash.go [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+// encodeHash extracts the first 40 bits of the hash from the hex string
+// (1 hex char represents 4 bits), and then maps vowels and vowel-like hex
+// characters to consonants to prevent bad words from being formed (the theory
+// is that no vowels makes it really hard to make bad words). Since the string
+// is hex, the only vowels it can contain are 'a' and 'e'.
+// We picked some arbitrary consonants to map to from the same character set as GenerateName.
+// See: https://github.com/kubernetes/apimachinery/blob/dc1f89aff9a7509782bde3b68824c8043a3e58cc/pkg/util/rand/rand.go#L75
+// If the hex string contains fewer than ten characters, returns an error.
+func encodeHash(hex string) (string, error) {
```
[the remark this answers is not in the document]
Comment (coldTea214) [reply pkg/webhook/default_server/sidecarset/mutating/hash.go]: I tested it, it's too long, e.g. {"test-sidecarset":"47c78v2x4777c67dvfzw6f5z467vd7bc4zf874wz6v5w7wdd8ffz48cwxz22czbx"}

File: pkg/controller/sidecarset/sidecarset_utils.go [hunk 17]
```diff
@@ -49,16 +58,17 @@ func calculateStatus(sidecarSet *appsv1alpha1.SidecarSet, pods []*corev1.Pod) (*
 }
 
 func isPodUpdated(sidecarSet *appsv1alpha1.SidecarSet, pod *corev1.Pod) (bool, error) {
```
[the remark this answers is not in the document]
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_utils.go]: done

File: pkg/controller/sidecarset/sidecarset_utils.go [hunk 18]
```diff
@@ -101,3 +112,218 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
...
+func (u *updatedPodCache) isUpdated(key, curHash string) bool {
+	u.lock.RLock()
+	hash := u.podSidecarUpdated[key]
+	u.lock.RUnlock()
+
+	if hash == curHash {
+		return true
+	}
+	if hash != "" && hash != curHash {
+		u.reset()
```
[the remark this answers is not in the document]
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_utils.go]: done

File: pkg/controller/sidecarset/sidecarset_controller.go [hunk 19]
```diff
@@ -131,7 +131,29 @@ func (r *ReconcileSidecarSet) Reconcile(request reconcile.Request) (reconcile.Re
...
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if sidecarSet.Spec.Paused {
+		klog.V(3).Infof("sidecarset %v is paused, skip update", sidecarSet.Name)
+		return reconcile.Result{}, nil
+	}
+
+	unavailableNum, err := getUnavailableNumber(sidecarSet, filteredPods)
```
[the remark this answers is not in the document]
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_controller.go]: done, see comment 'update procedure'

File: pkg/controller/sidecarset/sidecarset_utils.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_utils.go]: done

File: pkg/webhook/default_server/pod/mutating/pod_create_handler.go (hunk 14 above)
Comment (coldTea214) [reply pkg/webhook/default_server/pod/mutating/pod_create_handler.go]: done

File: pkg/controller/sidecarset/sidecarset_utils.go [hunk 20]
```diff
@@ -101,3 +109,204 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
...
+	if err := json.Unmarshal([]byte(pod.Annotations[hashKey]), &sidecarSetHash); err != nil {
+		return false, err
+	}
+	if _, ok := sidecarSetHash[sidecarSet.Name]; !ok {
+		return true, nil
+	}
+	return false, nil
+}
+
+// check are fields other than sidecar image had changed
```
Comment (Fei-Guo) [inline pkg/controller/sidecarset/sidecarset_utils.go]: check if fields other than sidecar image are changed.

File: pkg/controller/sidecarset/sidecarset_utils.go [hunk 21]
```diff
@@ -101,3 +109,204 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
...
+			sidecarSet.Annotations[sidecarsetmutating.SidecarSetHashAnnotation])
+		podInInformerUpdated, err := isPodSidecarUpdated(sidecarSet, pod)
+		if err != nil {
+			return 0, err
+		}
+		if podInCacheUpdated && !podInInformerUpdated {
+			unavailableNum++
+			continue
+		}
+
```
Comment (Fei-Guo) [inline pkg/controller/sidecarset/sidecarset_utils.go]: if podInCacheUpdated && podInInformerUpdated, should you remove the Pod entry from updateCache? In this way, once the Pod is confirmed being updated from informer, it will not consume the updateCache.

File: pkg/webhook/default_server/sidecarset/mutating/hash.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (Fei-Guo) [reply pkg/webhook/default_server/sidecarset/mutating/hash.go]: But will limiting the length of hash increase the chance of hash collision?

File: pkg/controller/sidecarset/sidecarset_controller.go [hunk 22]
```diff
@@ -131,7 +135,53 @@ func (r *ReconcileSidecarSet) Reconcile(request reconcile.Request) (reconcile.Re
...
 
-	// update sidecarset status
 	err = r.updateSidecarSetStatus(sidecarSet, status)
-	return reconcile.Result{}, err
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// update procedure:
+	// 1. check is sidecarset paused, if so, then quit
```
[the remark this answers is not in the document]
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_controller.go]: done

File: pkg/apis/apps/v1alpha1/sidecarset_types.go [hunk 23]
```diff
@@ -43,7 +46,7 @@ type SidecarSetStatus struct {
 	// SidecarSet's generation, which is updated on mutation by the API Server.
 	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
 
-	// matchedPods is the number of Pods whose labels are matched with this SidecarSet's selector
+	// matchedPods is the number of Pods whose labels are matched with this SidecarSet's selector and are created before sidecarset creates
```
[the remark this answers is not in the document]
Comment (coldTea214) [reply pkg/apis/apps/v1alpha1/sidecarset_types.go]: done

File: pkg/webhook/default_server/sidecarset/mutating/hash.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (coldTea214) [reply pkg/webhook/default_server/sidecarset/mutating/hash.go]: done

File: pkg/controller/sidecarset/sidecarset_utils.go (hunk 20 above)
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_utils.go]: done

File: pkg/controller/sidecarset/sidecarset_utils.go (hunk 21 above)
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_utils.go]: done

File: pkg/controller/sidecarset/sidecarset_utils.go [hunk 24]
```diff
@@ -101,3 +109,212 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
...
+func (u *updatedPodCache) isSidecarUpdated(key, curHash string) bool {
+	u.lock.RLock()
+	hash := u.podSidecarUpdated[key]
+	u.lock.RUnlock()
+
+	if hash == curHash {
+		return true
+	}
+	// if sidecarset changed, clean all stale cache
+	if hash != "" && hash != curHash {
```
Comment (xiang90) [inline pkg/controller/sidecarset/sidecarset_utils.go]: no need for hash!=curHash. there is a shortcut at line 163 for the equal case.

File: pkg/controller/sidecarset/sidecarset_utils.go [hunk 25]
```diff
@@ -101,3 +109,212 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
...
+		u.reset()
+		return false
```
Comment (xiang90) [inline pkg/controller/sidecarset/sidecarset_utils.go]: remove this line. it is unnecessary. it will return false at line 170 anyway.

File: pkg/controller/sidecarset/sidecarset_utils.go [hunk 26]
```diff
@@ -101,3 +109,212 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
...
+	var unavailableNum int
+	for _, pod := range pods {
+		// in case of informer cache latency
+		key := fmt.Sprintf("%v/%v/%v", sidecarSet.Name, pod.Namespace, pod.Name)
+		podInCacheUpdated := updateCache.isSidecarUpdated(key, sidecarSet.Annotations[sidecarsetmutating.SidecarSetHashAnnotation])
+		podInInformerUpdated, err := isPodSidecarUpdated(sidecarSet, pod)
+		if err != nil {
+			return 0, err
+		}
+		if podInCacheUpdated && !podInInformerUpdated {
```
Comment (xiang90) [inline pkg/controller/sidecarset/sidecarset_utils.go]: switch on case is probably better than a few if statements for this code block.

File: pkg/controller/sidecarset/sidecarset_utils.go (hunk 24 above)
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_utils.go]: it's true, but it will be easy to understand

File: pkg/controller/sidecarset/sidecarset_utils.go (hunk 26 above)
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_utils.go]: it might be easier to understand in this format

File: pkg/controller/sidecarset/sidecarset_utils.go (hunk 24 above)
Comment (coldTea214) [reply pkg/controller/sidecarset/sidecarset_utils.go]: done

**Comment by jian-he** [comment]
@coldTea214 
Do you have a UT for the updating procedure in the reconcile loop?

**Comment by coldTea214** [comment]
> @coldTea214
> Do you have a UT for the updating procedure in the reconcile loop? If not, could you add one ?

unit test added: pkg/controller/sidecarset/sidecarset_controller_test.go

**Comment by jian-he** [comment]
@coldTea214  btw, BroadcastJob has the [UT](https://github.com/openkruise/kruise/blob/c8b115118f849e8baf2e4d93ed5291dd03194ca4/pkg/controller/broadcastjob/broadcastjob_controller_test.go#L52)  that you may check. The underlying test framework might be useful for writing the UT
